content-services/content-sources-backend
# Fixes 2319: add repo filtering by origin & type

## Summary

This adds two new attributes to repositories:  content_type (rpm), and origin (external, red_hat).

I adjusted the values of 'origin' from the original task as the only other 'origin' i can think of would be for 'uploaded' content.  In the future when we support uploading content, we want image builder to still be able to get only 'custom repos with external urls', so they could fetch repos with origin=external to get that.

## Testing steps

For dev:

After migrating, run this main.go file:
```
package main

import (
	"github.com/content-services/content-sources-backend/pkg/db"
	"github.com/content-services/content-sources-backend/pkg/seeds"
	"github.com/openlyinc/pointy"
)

func main() {
	db.Connect()
	d := db.DB
	seeds.SeedRepositoryConfigurations(d, 1, seeds.SeedOptions{OrgID: "1", ContentType: pointy.Pointer("rpm"), Origin: pointy.Pointer("external")})
	seeds.SeedRepositoryConfigurations(d, 1, seeds.SeedOptions{OrgID: "1", ContentType: pointy.Pointer("not-rpm"), Origin: pointy.Pointer("red_hat")})
}
```

Launch the server, with the feature disabled:
```
FEATURES_NEW_REPO_FILTERING_ENABLED=false go run cmd/content-sources/main.go  api
```

Fetch the default repo list:

```
curl -X GET --location "http://localhost:8000/api/content-sources/v1.0/repositories/" \
    -H "Content-Type: application/json" \
    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiIxIn0sImFjY291bnRfbnVtYmVyIjoiMSIsImludGVybmFsIjp7Im9yZ19pZCI6IjEifX19Cg=="
```

This should show only 1 repo, as its filtering by to rpm & external by default.

Launch the server with the feature enabled:
```
FEATURES_NEW_REPO_FILTERING_ENABLED=true go run cmd/content-sources/main.go  api
```

Fetch the default repo list:

```
curl -X GET --location "http://localhost:8000/api/content-sources/v1.0/repositories/" \
    -H "Content-Type: application/json" \
    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiIxIn0sImFjY291bnRfbnVtYmVyIjoiMSIsImludGVybmFsIjp7Im9yZ19pZCI6IjEifX19Cg=="
```

Now you should get both repos that were created as part of the main.go run


Try adding  origin=external  & content_type=rpm to the url:

```
curl -X GET --location "http://localhost:8000/api/content-sources/v1.0/repositories/?origin=external" \
    -H "Content-Type: application/json" \
    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiIxIn0sImFjY291bnRfbnVtYmVyIjoiMSIsImludGVybmFsIjp7Im9yZ19pZCI6IjEifX19Cg=="
```

```
curl -X GET --location "http://localhost:8000/api/content-sources/v1.0/repositories/?content_type=rpm" \
    -H "Content-Type: application/json" \
    -H "x-rh-identity: eyJpZGVudGl0eSI6eyJ0eXBlIjoiVXNlciIsInVzZXIiOnsidXNlcm5hbWUiOiIxIn0sImFjY291bnRfbnVtYmVyIjoiMSIsImludGVybmFsIjp7Im9yZ19pZCI6IjEifX19Cg=="
```

You should get expected results.
Other things to test:

1.  trying to update the content_type or origin via PUT or PATCH should be ignored
2.  specifying content_type or origin when creating a repo should be ignored.
